`IntrinsicError: File exists (os error 17)` when attempting to overwrite a symlink in `dist/`

**Describe the bug**

If `pants export-codegen ...`'s is run on a target where the output includes a symlink, the _second_ attempt to write that to `dist/codegen` fails with `IntrinsicError: File exists (os error 17)`, potentially being unable to overwrite the symlink.

I'm guessing this applies to more than just `export-codegen` and `shell_command`, but that and `shell_command` are the easiest way I know to repeatedly attempt to write a symlink to a single path.

Reproducer:

```shell
cd $(mktemp -d)

cat > pants.toml <<EOF
[GLOBAL]
pants_version = "2.16.0rc1"
backend_packages = ["pants.backend.shell"]

[anonymous-telemetry]
enabled = false
EOF

cat > BUILD <<EOF
shell_command(
    name="symlink",
    command="ln -s whatever.txt link.txt",
    tools=["ln"],
    output_files=["link.txt"],
)

shell_command(
    name="normal",
    command="touch file.txt",
    tools=["touch"],
    output_files=["file.txt"],
)
EOF

pants export-codegen :symlink
# OK: ... Writing generated files to dist/codegen

pants export-codegen :symlink
# BUG: ... IntrinsicError: File exists (os error 17)

# delete the existing symlink, and try again:
rm dist/codegen/link.txt
pants export-codegen :symlink
# OK: ... Writing generated files to dist/codegen

# non-link works fine, for comparison:
pants export-codegen :normal
# OK: ... Writing generated files to dist/codegen

pants export-codegen :normal
# OK: ... Writing generated files to dist/codegen
```


All of the `export-codegen`s work, except for the `BUG` one:

```
16:42:31.28 [INFO] Writing generated files to dist/codegen
16:42:31.28 [ERROR] 1 Exception encountered:

Engine traceback:
  in `export-codegen` goal

IntrinsicError: File exists (os error 17)
```

**Pants version**
2.16.0rc1

**OS**
macOS

**Additional info**
https://github.com/pantsbuild/pants/issues/17758 is mildly related, since both that and this are pants failing to correctly update the contents of `dist/`.

--- a/src/rust/engine/fs/store/src/lib.rs
+++ b/src/rust/engine/fs/store/src/lib.rs
@@ -1442,12 +1442,28 @@ impl Store {
     destination: PathBuf,
     target: String,
   ) -> Result<(), StoreError> {
-    symlink(&target, &destination).await.map_err(|e| {
-      format!(
-        "Failed to create symlink to {target} at {}: {e}",
-        destination.display()
-      )
-    })?;
+    // Overwriting a symlink, even with another symlink, fails if it exists. This can occur when
+    // materializing to a fixed directory like dist/. To avoid pessimising the more common case (no
+    // overwrite, e.g. materializing to a temp dir), only remove after noticing a failure.
+    //
+    // NB. #17758, #18849: this is a work-around for inaccurate management of the contents of dist/.
+    for first in [true, false] {
+      match symlink(&target, &destination).await {
+        Ok(()) => break,
+        Err(e) if first && e.kind() == std::io::ErrorKind::AlreadyExists => {
+          tokio::fs::remove_dir_all(&destination).await.map_err(|e| {
+            format!(
+              "Failed to remove existing item at {} when creating symlink to {target} there: {e}",
+              destination.display()
+            )
+          })?
+        }
+        Err(e) => Err(format!(
+          "Failed to create symlink to {target} at {}: {e}",
+          destination.display()
+        ))?,
+      }
+    }
     Ok(())
   }
 

